### PR TITLE
Add user webhooks

### PR DESCRIPTION
**Pull request by KN4CK3R**
Currently we can add webhooks for organizations but not for users. This PR adds the latter. You can access it from the current users settings.

grafik

### CODE REVIEW

**Comment by lunny** [comment]
Please resolve conflicts.

File: routers/web/web.go [hunk 1]
```diff
@@ -862,7 +900,9 @@ func RegisterRoutes(m *web.Route) {
 				m.Post("/feishu/{id}", bindIgnErr(forms.NewFeishuHookForm{}), repo.FeishuHooksEditPost)
 				m.Post("/wechatwork/{id}", bindIgnErr(forms.NewWechatWorkHookForm{}), repo.WechatworkHooksEditPost)
 				m.Post("/packagist/{id}", bindIgnErr(forms.NewPackagistHookForm{}), repo.PackagistHooksEditPost)
-			}, webhooksEnabled)
+			}, webhooksEnabled, webhooksEnabled, func(ctx *context.Context) {
```
Comment (lunny) [inline routers/web/web.go]: duplicated?

File: routers/web/web.go [hunk 2]
```diff
@@ -736,7 +771,10 @@ func RegisterRoutes(m *web.Route) {
 					m.Post("/msteams/{id}", bindIgnErr(forms.NewMSTeamsHookForm{}), repo.MSTeamsHooksEditPost)
 					m.Post("/feishu/{id}", bindIgnErr(forms.NewFeishuHookForm{}), repo.FeishuHooksEditPost)
 					m.Post("/wechatwork/{id}", bindIgnErr(forms.NewWechatWorkHookForm{}), repo.WechatworkHooksEditPost)
-				}, webhooksEnabled)
+					m.Post("/packagist/{id}", bindIgnErr(forms.NewPackagistHookForm{}), repo.PackagistHooksEditPost)
+				}, webhooksEnabled, webhooksEnabled, func(ctx *context.Context) {
```
Comment (lunny) [inline routers/web/web.go]: duplicated?

**Comment by lunny** [comment]
Please resolve the conflicts

**Review by lunny — APPROVED** [review]

File: models/migrations/v1_19/v236.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+	if err := sess.Sync2(new(Webhook)); err != nil {
+		return err
+	}
+	if _, err := sess.Exec("UPDATE webhook SET owner_id = org_id"); err != nil {
+		return err
+	}
+	if err := base.DropTableColumns(sess, "webhook", "org_id"); err != nil {
+		return err
+	}
```
Comment (zeripath) [inline models/migrations/v1_19/v236.go]: Adjust this migration to look more like https://github.com/go-gitea/gitea/blob/main/models/migrations/v1_15/v184.go

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -808,6 +808,8 @@ remove_account_link = Remove Linked Account
 remove_account_link_desc = Removing a linked account will revoke its access to your Gitea account. Continue?
 remove_account_link_success = The linked account has been removed.
 
+hooks.desc = Add webhooks which will be triggered for <strong>all repositories</strong> under this user.
```
Comment (zeripath) [inline options/locale/locale_en-US.ini]:
```suggestion
hooks.desc = Add webhooks which will be triggered for <strong>all repositories</strong> owned by this user.
```

File: routers/api/v1/utils/hook.go [hunk 5]
```diff
@@ -18,15 +19,46 @@ import (
 	webhook_service "code.gitea.io/gitea/services/webhook"
 )
 
-// GetOrgHook get an organization's webhook. If there is an error, write to
+func ListOwnerHooks(ctx *context.APIContext, owner *user_model.User) {
```
Comment (zeripath) [inline routers/api/v1/utils/hook.go:23]:
```suggestion
// ListOwnerHooks lists the webhooks for a provided owner
func ListOwnerHooks(ctx *context.APIContext, owner *user_model.User) {
```

File: routers/api/v1/utils/hook.go [hunk 6]
```diff
@@ -197,21 +232,20 @@ func addHook(ctx *context.APIContext, form *api.CreateHookOption, orgID, repoID
 	return w, true
 }
 
-// EditOrgHook edit webhook `w` according to `form`. Writes to `ctx` accordingly
-func EditOrgHook(ctx *context.APIContext, form *api.EditHookOption, hookID int64) {
-	org := ctx.Org.Organization
-	hook, err := GetOrgHook(ctx, org.ID, hookID)
+// EditOwnerHook edit webhook `w` according to `form`. Writes to `ctx` accordingly
```
Comment (zeripath) [inline routers/api/v1/utils/hook.go]: This comment no longer matches this function

File: routers/api/v1/utils/hook.go [hunk 7]
```diff
@@ -325,3 +359,15 @@ func editHook(ctx *context.APIContext, form *api.EditHookOption, w *webhook.Webh
 	}
 	return true
 }
+
+func DeleteOwnerHook(ctx *context.APIContext, owner *user_model.User, hookID int64) {
```
Comment (zeripath) [inline routers/api/v1/utils/hook.go:401]: needs comment

File: routers/api/v1/user/hook.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
```
Comment (zeripath) [inline routers/api/v1/user/hook.go]:
```suggestion
// SPDX-License-Identifier: MIT
```

File: routers/web/user/setting/webhooks.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
```
Comment (zeripath) [inline routers/web/user/setting/webhooks.go]:
```suggestion
// SPDX-License-Identifier: MIT
```

File: templates/user/settings/navbar.tmpl [hunk 10]
```diff
@@ -23,6 +23,10 @@
 			{{.locale.Tr "packages.title"}}
 		</a>
 		{{end}}
+		{{if not DisableWebhooks}}
+		<a class="{{if .PageIsSettingsHooks}}active{{end}} item" href="{{AppSubUrl}}/user/settings/hooks">
+			{{.locale.Tr "repo.settings.hooks"}}
+		{{end}}
```
Comment (zeripath) [inline templates/user/settings/navbar.tmpl:33]:
```suggestion
		</a>
		{{end}}
```

File: routers/web/web.go [hunk 11]
```diff
@@ -474,6 +474,40 @@ func RegisterRoutes(m *web.Route) {
 		m.Get("/organization", user_setting.Organization)
 		m.Get("/repos", user_setting.Repos)
 		m.Post("/repos/unadopted", user_setting.AdoptOrDeleteRepository)
+
+		m.Group("/hooks", func() { //nolint:dupl
```
Comment (zeripath) [inline routers/web/web.go]: Perhaps we should extract this block out in to a separate function?

**Review by zeripath — APPROVED** [review]

**Comment by delvh** [comment]
Well…
You accidentally deleted the migration with the merge.

**Comment by lunny** [comment]
CI failure is related.

**Comment by jolheiser** [comment]
🎺 🤖

**Pull request merged**